scipy/optimize/_simulated_annealing.py

```diff
@@ -0,0 +1,559 @@
+            self.xmin = np.copy(x)
+            self.emin = e
+            self.not_improved_idx = 0
+            self.not_improved_max_idx = self.state.current_location.size
+            if e < self.state.ebest:
+                self.state.ebest = self.emin
+                self.state.xbest = np.copy(self.xmin)
+                self.state.current_energy = e
+                self.state.current_location = np.copy(x)
+
+
+class ObjectiveFunWrapper(object):
+    """
+    Class used to wrap around the objective function in order to apply local
+    search.
+    Default local minimizer is SciPy minimizer L-BFGS-B
+    """
+    
+    MAXITER_RATIO = 6
+    MAXITER_MIN = 100
+    MAXITER_MAX = 1000
+
+    def __init__(self, bounds, func, **kwargs):
+        self.func = func
+        self.nb_fun_call = 0
+        self.kwargs = kwargs
+        self.minimizer = minimize
+        self.fun_args = ()
+        bounds_list = list(zip(*bounds))
+        self.lower = np.array(bounds_list[0])
+        self.upper = np.array(bounds_list[1])
+        if 'args' in self.kwargs:
+            self.fun_args = self.kwargs.get('args')
+        
+    def set_default_minimizer(self):
+        # By default, use SciPy minimize with 'L-BFGS-B' method
+        n = len(self.lower)
+        ls_max_iter = min(max(n * self.MAXITER_RATIO, self.MAXITER_MIN),
+                          self.MAXITER_MAX)
+        self.kwargs['method'] = 'L-BFGS-B'
+        self.kwargs['options'] = {
+            'maxiter': ls_max_iter,
+        }
+        self.kwargs['bounds'] = list(zip(self.lower, self.upper))
+
+    def func_wrapper(self, x):
+        self.nb_fun_call += 1
+        return self.func(x, *self.fun_args)
+
+    def local_search(self, x):
+        x_tmp = np.copy(x)
+        fun_temp = self.func_wrapper(x)
+        mres = self.minimizer(self.func_wrapper, x, **self.kwargs)
+        # Check if is valid value
+        is_finite = np.all(np.isfinite(mres.x)) and np.isfinite(mres.fun)
+        in_bounds = np.all(mres.x >= self.lower) and np.all(
+            mres.x <= self.upper)
+        is_valid = is_finite and in_bounds
+
+        # Use the new point only if it is valid and return a better results
+        if is_valid and mres.fun < fun_temp:
+            return mres.fun, mres.x
+        else:
+            return fun_temp, x_tmp 
+
+
+def simulated_annealing(func, x0, bounds, maxiter=1000, local_search_options={},
+        initial_temp=5230., visit=2.62, accept=-5.0, maxfun=1e7, seed=None,
+        no_local_search=False):
+    """
+    Find the global minimum of a function using the Simulated Dual Annealing
+    algorithm
+
+    Parameters
+    ----------
+    func : callable
+        The objective function to be minimized.  Must be in the form
+        ``f(x, *args)``, where ``x`` is the argument in the form of a 1-D array
+        and ``args`` is a  tuple of any additional fixed parameters needed to
+        completely specify the function.
+    x0 : ndarray
```

Comment:
I have never studied simulated_annealing algorithm  but I have always thought it to be a population algorithm. So just a question, `x0` is a single initial condition point or a list of points?

This is not very clear from the documentation.

Besides that, I like it very much when the documentation specify the dimensions of the ndarray. Something like:
```
x0 : ndarray, shape (n, m)
    ...
bounds : sequence, shape (n,)
    ...
```

Reply:
x0 is the single optional initial coordinates for the algorithm to start. It can be used as a initial guess. I'll make this more clear in the documentation and specify dimensions. Thanks for this.